code/network/multimsgs.cpp

```diff
@@ -7611,6 +7580,220 @@ void process_NEW_primary_fired_packet(ubyte *data, header *hinfo)
 	}
 }
 
+const int NON_HOMING_PACKET_MISSILE	= (1 << 0);
+
+// Cyborg17 -- send the relative position of my fire to my current target as well as the frametime.
+// The whole goal of this is to create the weapon on the server that is heading at your target 
+// in the same exact way that it is on the client.
+void send_non_homing_fired_packet(ship* shipp, int banks_or_number_of_missiles_fired, bool secondary)
+{
+	Assertion(MULTIPLAYER_CLIENT, "Server is attempting to send the non_homing_fired_packet, which is client only. \n\nThis is a coder mistake.  Please report!");
+	if (MULTIPLAYER_MASTER) {
+		return;
+	}
+
+	int packet_size, objnum;
+	ubyte data[MAX_PACKET_SIZE], flags = 0; // ubanks_fired, current_bank;
+	object* objp;
+
+	// just in case nothing got fired
+	if (banks_or_number_of_missiles_fired <= 0) {
+		return;
+	}
+
+	// get an object pointer for this ship.
+	objnum = shipp->objnum;
+	objp = &Objects[objnum];
+
+	// I should never send non-homing packets for anyone except me
+	if (Player_obj != objp) {
+		return;
+	}
+
+	object* ref_objp = multi_get_network_object(multi_client_lookup_ref_obj_net_sig());
+	if (ref_objp == nullptr) {
+		mprintf(("Unable to get accurate reference object for non-homing packet.\n"));
+		if (!secondary) {
+			send_NEW_primary_fired_packet(shipp, banks_or_number_of_missiles_fired);
+		}
+		return;
+	}
+
+	BUILD_HEADER(LINEAR_WEAPON_FIRED);
+
+	if (secondary) {
+		flags |= NON_HOMING_PACKET_MISSILE;
+	}
+
+	// Add the shooting object.
+	ADD_USHORT(objp->net_signature);
+	ADD_DATA(flags);
+	ADD_USHORT(ref_objp->net_signature);
+
+	// We need the time elpased, so send the last frame we got from the server and how much time has happened since then.
+	int last_received_frame = multi_client_lookup_frame_idx();
+	auto time_elapsed = (ubyte)(timestamp() - multi_client_lookup_frame_timestamp());
```

Comment:
What happens if this wraps? Given sufficient lag it would be possible so if this would cause a problem then I'd suggest making this a ushort.

Reply:
Changes made....